Cython/Compiler/ModuleNode.py

```diff
@@ -1048,6 +1049,7 @@ def generate_cpp_class_definition(self, entry, code):
                     [base_class.empty_declaration_code() for base_class in type.base_classes])
                 code.put(" : public %s" % base_class_decl)
             code.putln(" {")
+            self.generate_type_header_code(scope.type_entries, code)
```

Comment:
Here we are in the C++ code generation part.

Just added the correct method to generate the C++ code for the types defined inside a given cppclass in the cython code.

C++ code generation wise, this is in fact the main intent of this PR, the rest is more bug fix than feature addition.